app/src/main/java/pm/gnosis/heimdall/ui/safe/main/SafeMainActivity.kt

```diff
@@ -405,17 +414,26 @@ class SafeMainActivity : ViewModelActivity<SafeMainContract>() {
         popupMenu.menu.findItem(R.id.safe_details_menu_connect).isVisible = false
     }
 
-    private fun isConnectedToBrowserExtension(isConnected: Boolean) {
+    @ExperimentalCoroutinesApi
+    private fun handleSafeConfig(newMasterCopy: Solidity.Address?, isConnected: Boolean) {
         isConnectedToExtension = isConnected
+        val safe = safeSubject.value as? Safe
+        val canUpgrade = safe != null && newMasterCopy != null
+        layout_safe_main_upgrade_warning_container.visible(canUpgrade)
+        layout_safe_main_menu_upgrade_warning.visible(canUpgrade)
+        if (canUpgrade) {
```

Comment:
are edgecases possible when upgrade cannot be made? How do we handle them?

Reply:
I am not aware of any edge cases. Do you have anything in mind?

Else they should be handled on the screen where we build the upgrade tx in my opinion

Reply:
maybe incompatible contract versions if that's possible. Just speculations